File: src/scpn_phase_orchestrator/oscillators/quality.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Phase quality scorer

"""Quality aggregation and collapse detection for extracted phase states.

The scorer turns per-oscillator extraction quality into weighted aggregate
signals for runtime gating and diagnostics. Empty state sets collapse to safe
defaults, low-quality states can be masked, and amplitude weighting prevents
near-zero signals from dominating quality summaries.
"""

from __future__ import annotations

from typing import TypeAlias

import numpy as np
from numpy.typing import NDArray

from scpn_phase_orchestrator.oscillators.base import PhaseState

__all__ = ["PhaseQualityScorer"]

FloatArray: TypeAlias = NDArray[np.float64]

try:
    from spo_kernel import PyPhaseQualityScorer as _RustPhaseQualityScorer
except ImportError:
    _RustPhaseQualityScorer = None


class PhaseQualityScorer:
    """Aggregate quality scoring and collapse detection for phase state arrays."""

    def __init__(self, collapse_threshold: float = 0.1, min_quality: float = 0.3):
        if not np.isfinite(collapse_threshold):
            raise ValueError("collapse_threshold must be finite")
        if not np.isfinite(min_quality):
            raise ValueError("min_quality must be finite")
        if not 0.0 <= collapse_threshold <= 1.0:
            raise ValueError("collapse_threshold must be in [0, 1]")
        if not 0.0 <= min_quality <= 1.0:
            raise ValueError("min_quality must be in [0, 1]")
        self._collapse_threshold = float(collapse_threshold)
        self._min_quality = float(min_quality)
        self._rust = (
            _RustPhaseQualityScorer(
                collapse_threshold=self._collapse_threshold,
                min_quality=self._min_quality,
            )
            if _RustPhaseQualityScorer is not None
            else None
        )

    def score(self, phase_states: list[PhaseState]) -> float:
        """Weighted average quality across all phase states."""
        if not phase_states:
            return 0.0
        qualities = np.array([ps.quality for ps in phase_states])
        amplitudes = np.array([ps.amplitude for ps in phase_states])
        if self._rust is not None:
            return float(self._rust.score(qualities.tolist(), amplitudes.tolist()))
        weights = np.maximum(amplitudes, 1e-12)
        return float(np.average(qualities, weights=weights))

    # Thresholds: see docs/ASSUMPTIONS.md § Quality Gating
    def detect_collapse(
        self, phase_states: list[PhaseState], threshold: float = 0.1
    ) -> bool:
        """Return True if quality is below threshold for the majority of states."""
        if not phase_states:
            return True
        if not np.isfinite(threshold):
            raise ValueError("threshold must be finite")
        threshold = float(threshold)
        if not 0.0 <= threshold <= 1.0:
            raise ValueError("threshold must be in [0, 1]")
        if self._rust is not None and threshold == self._collapse_threshold:
            qualities = [ps.quality for ps in phase_states]
            return bool(self._rust.is_collapsed(qualities))
        below = sum(1 for ps in phase_states if ps.quality < threshold)
        return below > len(phase_states) / 2

    def downweight_mask(
        self, phase_states: list[PhaseState], min_quality: float = 0.3
    ) -> FloatArray:
        """Weight array in [0,1], zeros below min_quality."""
        if not phase_states:
            return np.array([], dtype=np.float64)
        if not np.isfinite(min_quality):
            raise ValueError("min_quality must be finite")
        min_quality = float(min_quality)
        if not 0.0 <= min_quality <= 1.0:
            raise ValueError("min_quality must be in [0, 1]")
        qualities = np.array([ps.quality for ps in phase_states])
        if self._rust is not None and min_quality == self._min_quality:
            return np.asarray(
                self._rust.downweight_mask(qualities.tolist()),
                dtype=np.float64,
            )
        mask = np.where(qualities >= min_quality, qualities, 0.0)
        return mask.astype(np.float64)

File: spo-kernel/crates/spo-oscillators/src/quality.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Phase Orchestrator — Phase quality scorer

/// Aggregate quality scoring and collapse detection.
#[derive(Clone, Debug)]
pub struct PhaseQualityScorer {
    pub collapse_threshold: f64,
    pub min_quality: f64,
}

impl Default for PhaseQualityScorer {
    fn default() -> Self {
        Self {
            collapse_threshold: 0.1,
            min_quality: 0.3,
        }
    }
}

impl PhaseQualityScorer {
    /// Weighted average quality. `qualities` and `amplitudes` must be same length.
    ///
    /// # Arguments
    ///
    /// * `qualities` - Per-oscillator quality values in [0, 1].
    /// * `amplitudes` - Per-oscillator amplitudes used as weights.
    #[must_use]
    pub fn score(&self, qualities: &[f64], amplitudes: &[f64]) -> f64 {
        if qualities.is_empty() {
            return 0.0;
        }
        if !self.collapse_threshold.is_finite() || !self.min_quality.is_finite() {
            return 0.0;
        }
        let n = qualities.len().min(amplitudes.len());
        let (wsum, total_w) = (0..n).fold((0.0, 0.0), |(ws, tw), i| {
            let q = qualities[i];
            let amp = amplitudes[i];
            if !q.is_finite() || !amp.is_finite() {
                return (ws, tw);
            }
            let q = q.clamp(0.0, 1.0);
            let w = amp.max(1e-12);
            (ws + q * w, tw + w)
        });
        if total_w <= 0.0 {
            return 0.0;
        }
        wsum / total_w
    }

    /// True if quality is below threshold for the majority of states.
    #[must_use]
    pub fn is_collapsed(&self, qualities: &[f64]) -> bool {
        if qualities.is_empty() {
            return true;
        }
        if !self.collapse_threshold.is_finite() {
            return true;
        }
        let below = qualities
            .iter()
            .filter(|&&q| !q.is_finite() || q < self.collapse_threshold)
            .count();
        below > qualities.len() / 2
    }

    /// Weight array: qualities above min_quality pass through, others zeroed.
    #[must_use]
    pub fn downweight_mask(&self, qualities: &[f64]) -> Vec<f64> {
        if !self.min_quality.is_finite() {
            return vec![0.0; qualities.len()];
        }
        qualities
            .iter()
            .map(|&q| {
                if q.is_finite() && q >= self.min_quality {
                    q.clamp(0.0, 1.0)
                } else {
                    0.0
                }
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn score_uniform() {
        let s = PhaseQualityScorer::default();
        let q = vec![0.8, 0.8, 0.8];
        let a = vec![1.0, 1.0, 1.0];
        assert!((s.score(&q, &a) - 0.8).abs() < 1e-12);
    }

    #[test]
    fn score_weighted() {
        let s = PhaseQualityScorer::default();
        let q = vec![1.0, 0.0];
        let a = vec![2.0, 1.0];
        // (1.0*2 + 0.0*1) / (2+1) = 2/3
        assert!((s.score(&q, &a) - 2.0 / 3.0).abs() < 1e-12);
    }

    #[test]
    fn score_empty() {
        assert_eq!(PhaseQualityScorer::default().score(&[], &[]), 0.0);
    }

    #[test]
    fn collapse_all_low() {
        let s = PhaseQualityScorer::default();
        assert!(s.is_collapsed(&[0.01, 0.02, 0.03]));
    }

    #[test]
    fn collapse_all_high() {
        let s = PhaseQualityScorer::default();
        assert!(!s.is_collapsed(&[0.8, 0.9, 0.7]));
    }

    #[test]
    fn collapse_empty() {
        assert!(PhaseQualityScorer::default().is_collapsed(&[]));
    }

    #[test]
    fn downweight_mask_filters() {
        let s = PhaseQualityScorer::default();
        let mask = s.downweight_mask(&[0.1, 0.5, 0.3, 0.9]);
        assert_eq!(mask[0], 0.0);
        assert_eq!(mask[1], 0.5);
        assert_eq!(mask[2], 0.3);
        assert_eq!(mask[3], 0.9);
    }

    #[test]
    fn score_ignores_non_finite_inputs() {
        let s = PhaseQualityScorer::default();
        let q = vec![0.8, f64::NAN, 0.2];
        let a = vec![1.0, 1.0, f64::INFINITY];
        assert!((s.score(&q, &a) - 0.8).abs() < 1e-12);
    }

    #[test]
    fn non_finite_thresholds_fail_closed() {
        let s = PhaseQualityScorer {
            collapse_threshold: f64::NAN,
            min_quality: f64::INFINITY,
        };
        assert_eq!(s.score(&[0.5], &[1.0]), 0.0);
        assert!(s.is_collapsed(&[0.9, 0.8]));
        assert_eq!(s.downweight_mask(&[0.9, 0.8]), vec![0.0, 0.0]);
    }
}
